core/tests/vm-benchmark/benches/instructions.rs

```diff
@@ -0,0 +1,198 @@
+//! Measures instruction counts for the fuzzed bytecodes.
+
+use std::{
+    collections::BTreeMap,
+    env,
+    sync::{Arc, Mutex},
+};
+
+use vise::{Gauge, LabeledFamily, Metrics};
+use vm_benchmark::{
+    criterion::PrometheusRuntime, BenchmarkingVm, BenchmarkingVmFactory, Fast, Legacy, BYTECODES,
+};
+use yab::{
+    reporter::{BenchmarkOutput, BenchmarkReporter, Reporter},
+    AccessSummary, Bencher, BenchmarkId,
+};
+
+fn benchmarks_for_vm<VM: BenchmarkingVmFactory>(bencher: &mut Bencher) {
+    bencher.bench(
+        BenchmarkId::new("init", VM::LABEL.as_str()),
+        BenchmarkingVm::<VM>::default,
+    );
+
+    for bytecode in BYTECODES {
+        bencher.bench_with_capture(
+            BenchmarkId::new(bytecode.name, VM::LABEL.as_str()),
+            |capture| {
+                let mut vm = yab::black_box(BenchmarkingVm::<VM>::default());
+                let tx = yab::black_box(bytecode.deploy_tx());
+                capture.measure(|| vm.run_transaction(&tx));
+            },
+        );
+    }
+}
+
+/// Reporter that pushes cachegrind metrics to Prometheus.
+#[derive(Debug)]
+struct MetricsReporter {
+    _runtime: Option<PrometheusRuntime>,
+}
+
+impl Default for MetricsReporter {
+    fn default() -> Self {
+        Self {
+            _runtime: PrometheusRuntime::new(),
+        }
+    }
+}
+
+impl Reporter for MetricsReporter {
+    fn new_benchmark(&mut self, id: &BenchmarkId) -> Box<dyn BenchmarkReporter> {
+        Box::new(MetricsBenchmarkReporter(id.clone()))
+    }
+}
+
+#[derive(Debug)]
+struct MetricsBenchmarkReporter(BenchmarkId);
+
+impl BenchmarkReporter for MetricsBenchmarkReporter {
+    fn ok(self: Box<Self>, output: &BenchmarkOutput) {
+        #[derive(Debug, Metrics)]
+        #[metrics(prefix = "vm_cachegrind")]
+        struct VmCachegrindMetrics {
+            #[metrics(labels = ["benchmark"])]
+            instructions: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l1_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            l2_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            ram_accesses: LabeledFamily<String, Gauge<u64>>,
+            #[metrics(labels = ["benchmark"])]
+            cycles: LabeledFamily<String, Gauge<u64>>,
+        }
+
+        #[vise::register]
+        static VM_CACHEGRIND_METRICS: vise::Global<VmCachegrindMetrics> = vise::Global::new();
+
+        let id = self.0.to_string();
+        VM_CACHEGRIND_METRICS.instructions[&id].set(output.stats.total_instructions());
+        if let Some(&full) = output.stats.as_full() {
+            let summary = AccessSummary::from(full);
+            VM_CACHEGRIND_METRICS.l1_accesses[&id].set(summary.l1_hits);
+            VM_CACHEGRIND_METRICS.l2_accesses[&id].set(summary.l3_hits);
```

Comment:
Typo or inconsistent naming?

Reply:
Is Cachegrind's L1 a 4-cycle latency cache or more like L2?

Reply:
AFAICT, it's intended to be L1 cache; see [this article](https://pythonspeed.com/articles/consistent-benchmarking-in-ci/) and [valgrind docs](https://valgrind.org/docs/manual/cg-manual.html).